### PR TITLE
Add Indoor/Outdoor Meter

### PR DESCRIPTION
**Pull request by boozer2**
Add support for "Indoor/Outdoor Thermo-Hygrometer" (aka Indoor/Outdoor Meter).

### CODE REVIEW

**Review by bdraco — CHANGES_REQUESTED** [review]
Please add coverage for the new sensor and address any test failures.

Thanks

**Comment by Pierre-33** [comment]
Hey! I'm very interested by this PR request and willing to help :)
I'm not sure why this if was added in the first place:
```python
if not temp_data: 
   temp_data = data[3:6]
```
but it would seems that when mfr_data is None, temp_data is not set before reaching this part of the code.
A simple 
```python
temp_data = None
```
at the start of the function should fix the test.

**Comment by boozer2** [comment]
> Hey! I'm very interested by this PR request and willing to help :)

Thanks! I did already figure out the fix for that issue, but I haven't done the coverage for the new sensor yet, and I'm not sure how to initiate those tests myself. I'll try and push some updates this weekend with both the fix and new coverage tests.

I did already test my original updates and despite the obvious coding error, both my old Meter and new I/O Meter worked great in HA. Just took a small addition to the HA integration code for the new sensor.

I have about 10 of these now so I'd also like to get them supported!

**Comment by Pierre-33** [comment]
That's great news! :) I only have a couple of them and I use an esphome node to fetch their reading. Did you have any success retrieving the other data (Dew points and the like)?

Unfortunately, this coverage thing is quite obscure for me as well :-/

**Comment by boozer2** [comment]
> Did you have any success retrieving the other data (Dew points and the like)?

I don't think this information is advertised by the device. I get the impression it's calculated by the app using the temp/humidity data. 

These online calculators will allow you to get these values using just temp and relative humidity (I get results very close to what the SwitchBot app shows, just make sure the selected units match what the app uses):
https://www.omnicalculator.com/physics/absolute-humidity
https://www.omnicalculator.com/physics/dew-point
https://www.omnicalculator.com/biology/vapor-pressure-deficit

I hope to finish coverage and push the new changes in the next day or two.

**Comment by bdraco** [comment]
CI passed. Do you have a Home Assistant PR as well ?

I bought one of these to test this but it would be a lot easier if you already have a draft HA PR

**Comment by boozer2** [comment]
> CI passed. Do you have a Home Assistant PR as well ?

I didn't submit PR there yet. All that's required is a one-line change to const.py in the integration within:
`NON_CONNECTABLE_SUPPORTED_MODEL_TYPES`
to add:
`    SwitchbotModel.IO_METER: SupportedModels.HYGROMETER,`

I can work on this but I probably need to rebase to submit the PR and to be honest I'm far from an experienced coder and I'm still learning git... I'll take a look later tonight or tomorrow.

**Comment by boozer2** [comment]
@bdraco I added a draft PR to homeassistant to add the I/O Meter. Let me know if I need to do anything else.

https://github.com/home-assistant/core/pull/94836

File: switchbot/adv_parser.py [hunk 1]
```diff
@@ -93,7 +99,6 @@ class SwitchbotSupportedType(TypedDict):
         "modelName": SwitchbotModel.PLUG_MINI,
         "modelFriendlyName": "Plug Mini",
         "func": process_woplugmini,
-        "manufacturer_data_length": 12,
```
Comment (bdraco) [inline switchbot/adv_parser.py:96]: The plug has the same length as the iometer so we had to drop this to get detection working

**Review by bdraco — APPROVED** [review]

**Comment by Nardol** [comment]
Sorry for commenting a merged PR but @Pierre-33 could you provide the code for indoor/outdoor meters you use with ESPHome please?

**Comment by Pierre-33** [comment]
Sorry, I deleted my ESPHome config a few days ago now that the indoor/outdoor thermometers are supported by the official HA integration (thanks again for that btw!).

I didn't know about Bluetooth Proxy in HA until very recently, so maybe you don't know about it as well :)
You just have to flash an esp32 with that : https://esphome.github.io/bluetooth-proxies/ and your thermometer should show up in Home Assistant.

If for any reason you still want to go through ESPHome, I started with the code found on that [thread ](https://forum.hacf.fr/t/recuperer-la-temperature-lhygrometrie-et-le-niveau-de-batterie-de-ses-thermometres-switch-bot-avec-esphome/5671)and maybe did a couple of adjustments in the Lambda based on that PR request. I'm not sure it's been a while ago now :-/ Unfortunately the thread is in French, but the code part is at the very beginning of it.

**Comment by Nardol** [comment]
> I didn't know about Bluetooth Proxy in HA until very recently, so maybe you don't know about it as well :) You just have to flash an esp32 with that : https://esphome.github.io/bluetooth-proxies/ and your thermometer should show up in Home Assistant.
> 

I needed to know what you did because a friend of mine is going to receive one of these device and he does not use HA at all.
Personally I use the method you describe, which works well.

> If for any reason you still want to go through ESPHome, I started with the code found on that [thread ](https://forum.hacf.fr/t/recuperer-la-temperature-lhygrometrie-et-le-niveau-de-batterie-de-ses-thermometres-switch-bot-avec-esphome/5671)and maybe did a couple of adjustments in the Lambda based on that PR request. I'm not sure it's been a while ago now :-/ Unfortunately the thread is in French, but the code part is at the very beginning of it.

Oh nice, my tutorial :smile: and the modification you made is the part which interests me :slightly_smiling_face: I wanted to anticipate but I will see how it will work and try to modify within basing to this PR, thanks.

**Comment by Pierre-33** [comment]
Such a small world! :) Thanks for the tutorial then, it's been really helpful, and that's actually I discovered Bluetooth proxy

Actually, the trick was to fetch the data from the manufacturer data, that should put you on the right track :

```yaml
esp32_ble_tracker:
  on_ble_advertise:
    - mac_address: E5:65:1B:86:06:FB
      then:
        - lambda: |-            
            for (auto data : x.get_manufacturer_datas()) 
            {
                if(data.data.size() == 12) 
                {
                  float temperature = (float)(data.data[9] & 0b01111111) + ((float)(data.data[8] & 0b00001111) / 10);
                  if (!(data.data[9] & 0b10000000)) 
                  {
                    temperature = -temperature;
                  }
                  int8_t humidity= data.data[10] & 0b01111111;

                  ESP_LOGD("ble_adv", "Temperature : %.2f°C", temperature);
                  ESP_LOGD("ble_adv", "Humidity : %d%%", humidity);
                }
            }
            for (auto data : x.get_service_datas()) 
            {
              if(data.data.size() == 3)
              {
                // Not so sure about the battery
                int8_t battery = data.data[2] & 0b01111111;
                ESP_LOGD("ble_adv", "Battery : %d%%", battery);
              } 
            }
```

**Comment by Nardol** [comment]
Thanks a lot! And now I am sure it is not possible to get battery level using passive scan if it is only possible using service data, not manufacturer. Except if I miss-understood something as I never succeeded in understanding anything in the official BLE API docs from Switch Bot :laughing:

**Comment by Pierre-33** [comment]
Don't assume anything from my code :) I didn't read any doc and didn't know anything about Bluetooth. I just notice that service data on that sensor were too small and give it a try with manufacturer data and the data offset present in that lib. Turns out it works, doesn't mean there is no better way to do this :)

**Pull request merged**